diff --git a/fastboot/fastboot.cpp b/fastboot/fastboot.cpp
index d9050931..66385be9 100644
--- a/fastboot/fastboot.cpp
+++ b/fastboot/fastboot.cpp
@@ -41,13 +41,13 @@ Device::BuiltinAction StartFastboot(Device* device, const std::vector<std::strin
 
   std::vector<std::string> title_lines;
   title_lines.push_back("Product name - " + android::base::GetProperty("ro.product.device", ""));
-  title_lines.push_back("Bootloader version - " + android::base::GetProperty("ro.bootloader", ""));
-  title_lines.push_back("Baseband version - " +
-                        android::base::GetProperty("ro.build.expect.baseband", ""));
+  //title_lines.push_back("Bootloader version - " + android::base::GetProperty("ro.bootloader", ""));
+  //title_lines.push_back("Baseband version - " +
+  //                      android::base::GetProperty("ro.build.expect.baseband", ""));
   title_lines.push_back("Serial number - " + android::base::GetProperty("ro.serialno", ""));
   title_lines.push_back(std::string("Secure boot - ") +
                         ((android::base::GetProperty("ro.secure", "") == "1") ? "yes" : "no"));
-  title_lines.push_back("HW version - " + android::base::GetProperty("ro.revision", ""));
+  //title_lines.push_back("HW version - " + android::base::GetProperty("ro.revision", ""));
 
   ui->ResetKeyInterruptStatus();
   ui->SetTitle(title_lines);
