diff --git a/build/soong/Android.bp b/build/soong/Android.bp
index 382b68ee..1eb41ae1 100644
--- a/build/soong/Android.bp
+++ b/build/soong/Android.bp
@@ -28,14 +28,15 @@ lineage_generator {
     name: "generated_kernel_includes",
 
     // The headers make command
-    cmd: "make $(KERNEL_MAKE_FLAGS) -C $(TARGET_KERNEL_SOURCE) O=$(genDir) ARCH=$(KERNEL_ARCH) $(KERNEL_CROSS_COMPILE) headers_install",
+    //cmd: "make $(KERNEL_MAKE_FLAGS) -C $(TARGET_KERNEL_SOURCE) O=$(genDir) ARCH=$(KERNEL_ARCH) $(KERNEL_CROSS_COMPILE) headers_install",
+    cmd: "echo 'Skip generated kernel includes'",
 
     // Directories that can be imported by a cc_* module generated_headers property
     export_include_dirs: ["usr/include", "usr/techpack/audio/include"],
 
     // Sources for dependency tracking
-    dep_root: "$(TARGET_KERNEL_SOURCE)",
-    dep_files: [ "Makefile", "include/**/*", "arch/$(KERNEL_ARCH)/include/**/*", "techpack/audio/include/**/*"],
+    //dep_root: "$(TARGET_KERNEL_SOURCE)",
+    //dep_files: [ "Makefile", "include/**/*", "arch/$(KERNEL_ARCH)/include/**/*", "techpack/audio/include/**/*"],
 }
 
 cc_library_headers {
