diff --git a/core/Makefile b/core/Makefile
index 027df5828..b98bc1b99 100644
--- a/core/Makefile
+++ b/core/Makefile
@@ -3473,11 +3473,13 @@ endif # end of AVB for BUILDING_SYSTEM_OTHER_IMAGE
 
 INTERNAL_AVB_PARTITIONS_IN_CHAINED_VBMETA_IMAGES := \
     $(BOARD_AVB_VBMETA_SYSTEM) \
+    $(BOARD_AVB_VBMETA_SYSTEM_EXT) \
+    $(BOARD_AVB_VBMETA_PRODUCT) \
     $(BOARD_AVB_VBMETA_VENDOR)
 
 # Not allowing the same partition to appear in multiple groups.
 ifneq ($(words $(sort $(INTERNAL_AVB_PARTITIONS_IN_CHAINED_VBMETA_IMAGES))),$(words $(INTERNAL_AVB_PARTITIONS_IN_CHAINED_VBMETA_IMAGES)))
-  $(error BOARD_AVB_VBMETA_SYSTEM and BOARD_AVB_VBMETA_VENDOR cannot have duplicates)
+  $(error BOARD_AVB_VBMETA_SYSTEM, BOARD_AVB_VBMETA_SYSTEM_EXT, BOARD_AVB_VBMETA_PRODUCT, and BOARD_AVB_VBMETA_VENDOR cannot have duplicates)
 endif
 
 # When building a standalone recovery image for non-A/B devices, recovery image must be self-signed
@@ -3661,6 +3663,14 @@ ifdef BOARD_AVB_VBMETA_SYSTEM
 $(eval $(call check-and-set-avb-args,vbmeta_system))
 endif
 
+ifdef BOARD_AVB_VBMETA_SYSTEM_EXT
+$(eval $(call check-and-set-avb-args,vbmeta_system_ext))
+endif
+
+ifdef BOARD_AVB_VBMETA_PRODUCT
+$(eval $(call check-and-set-avb-args,vbmeta_product))
+endif
+
 ifdef BOARD_AVB_VBMETA_VENDOR
 $(eval $(call check-and-set-avb-args,vbmeta_vendor))
 endif
@@ -3685,6 +3695,8 @@ endif
 BOARD_AVB_MAKE_VBMETA_IMAGE_ARGS += --padding_size 4096
 BOARD_AVB_MAKE_VBMETA_SYSTEM_IMAGE_ARGS += --padding_size 4096
 BOARD_AVB_MAKE_VBMETA_VENDOR_IMAGE_ARGS += --padding_size 4096
+BOARD_AVB_MAKE_VBMETA_SYSTEM_EXT_IMAGE_ARGS += --padding_size 4096
+BOARD_AVB_MAKE_VBMETA_PRODUCT_IMAGE_ARGS += --padding_size 4096
 
 ifeq (eng,$(filter eng, $(TARGET_BUILD_VARIANT)))
 # We only need the flag in top-level vbmeta.img.
@@ -3700,6 +3712,16 @@ BOARD_AVB_MAKE_VBMETA_SYSTEM_IMAGE_ARGS += \
     --rollback_index $(BOARD_AVB_VBMETA_SYSTEM_ROLLBACK_INDEX)
 endif
 
+ifdef BOARD_AVB_VBMETA_SYSTEM_EXT_ROLLBACK_INDEX
+BOARD_AVB_MAKE_VBMETA_SYSTEM_EXT_IMAGE_ARGS += \
+    --rollback_index $(BOARD_AVB_VBMETA_SYSTEM_EXT_ROLLBACK_INDEX)
+endif
+
+ifdef BOARD_AVB_VBMETA_PRODUCT_ROLLBACK_INDEX
+BOARD_AVB_MAKE_VBMETA_PRODUCT_IMAGE_ARGS += \
+    --rollback_index $(BOARD_AVB_VBMETA_PRODUCT_ROLLBACK_INDEX)
+endif
+
 ifdef BOARD_AVB_VBMETA_VENDOR_ROLLBACK_INDEX
 BOARD_AVB_MAKE_VBMETA_VENDOR_IMAGE_ARGS += \
     --rollback_index $(BOARD_AVB_VBMETA_VENDOR_ROLLBACK_INDEX)
@@ -3737,6 +3759,12 @@ define extract-avb-chain-public-keys
   $(if $(BOARD_AVB_VBMETA_SYSTEM_KEY_PATH),\
     $(hide) $(AVBTOOL) extract_public_key --key $(BOARD_AVB_VBMETA_SYSTEM_KEY_PATH) \
         --output $(1)/vbmeta_system.avbpubkey)
+  $(if $(BOARD_AVB_VBMETA_SYSTEM_EXT_KEY_PATH),\
+    $(hide) $(AVBTOOL) extract_public_key --key $(BOARD_AVB_VBMETA_SYSTEM_EXT_KEY_PATH) \
+        --output $(1)/vbmeta_system_ext.avbpubkey)
+  $(if $(BOARD_AVB_VBMETA_PRODUCT_KEY_PATH),\
+    $(hide) $(AVBTOOL) extract_public_key --key $(BOARD_AVB_VBMETA_PRODUCT_KEY_PATH) \
+        --output $(1)/vbmeta_product.avbpubkey)
   $(if $(BOARD_AVB_VBMETA_VENDOR_KEY_PATH),\
     $(hide) $(AVBTOOL) extract_public_key --key $(BOARD_AVB_VBMETA_VENDOR_KEY_PATH) \
         --output $(1)/vbmeta_vendor.avbpubkey)
@@ -3775,6 +3803,24 @@ $(INSTALLED_VBMETA_SYSTEMIMAGE_TARGET): \
 endif
 endif # BUILDING_SYSTEM_IMAGE
 
+ifdef BOARD_AVB_VBMETA_SYSTEM_EXT
+INSTALLED_VBMETA_SYSTEM_EXTIMAGE_TARGET := $(PRODUCT_OUT)/vbmeta_system_ext.img
+$(INSTALLED_VBMETA_SYSTEM_EXTIMAGE_TARGET): \
+	    $(AVBTOOL) \
+	    $(call images-for-partitions,$(BOARD_AVB_VBMETA_SYSTEM_EXT)) \
+	    $(BOARD_AVB_VBMETA_SYSTEM_EXT_KEY_PATH)
+	$(call build-chained-vbmeta-image,vbmeta_system_ext)
+endif
+
+ifdef BOARD_AVB_VBMETA_PRODUCT
+INSTALLED_VBMETA_PRODUCTIMAGE_TARGET := $(PRODUCT_OUT)/vbmeta_product.img
+$(INSTALLED_VBMETA_PRODUCTIMAGE_TARGET): \
+	    $(AVBTOOL) \
+	    $(call images-for-partitions,$(BOARD_AVB_VBMETA_PRODUCT)) \
+	    $(BOARD_AVB_VBMETA_PRODUCT_KEY_PATH)
+	$(call build-chained-vbmeta-image,vbmeta_product)
+endif
+
 ifdef BOARD_AVB_VBMETA_VENDOR
 INSTALLED_VBMETA_VENDORIMAGE_TARGET := $(PRODUCT_OUT)/vbmeta_vendor.img
 $(INSTALLED_VBMETA_VENDORIMAGE_TARGET): \
@@ -3814,8 +3860,12 @@ $(INSTALLED_VBMETAIMAGE_TARGET): \
 	    $(INSTALLED_CUSTOMIMAGES_TARGET) \
 	    $(INSTALLED_RECOVERYIMAGE_TARGET) \
 	    $(INSTALLED_VBMETA_SYSTEMIMAGE_TARGET) \
+	    $(INSTALLED_VBMETA_SYSTEM_EXTIMAGE_TARGET) \
+	    $(INSTALLED_VBMETA_PRODUCTIMAGE_TARGET) \
 	    $(INSTALLED_VBMETA_VENDORIMAGE_TARGET) \
 	    $(BOARD_AVB_VBMETA_SYSTEM_KEY_PATH) \
+	    $(BOARD_AVB_VBMETA_SYSTEM_EXT_KEY_PATH) \
+	    $(BOARD_AVB_VBMETA_PRODUCT_KEY_PATH) \
 	    $(BOARD_AVB_VBMETA_VENDOR_KEY_PATH) \
 	    $(BOARD_AVB_KEY_PATH)
 	$(build-vbmetaimage-target)
@@ -4394,6 +4444,20 @@ ifneq (,$(strip $(BOARD_AVB_VBMETA_SYSTEM)))
 	$(hide) echo "avb_vbmeta_system_algorithm=$(BOARD_AVB_VBMETA_SYSTEM_ALGORITHM)" >> $@
 	$(hide) echo "avb_vbmeta_system_rollback_index_location=$(BOARD_AVB_VBMETA_SYSTEM_ROLLBACK_INDEX_LOCATION)" >> $@
 endif # BOARD_AVB_VBMETA_SYSTEM
+ifneq (,$(strip $(BOARD_AVB_VBMETA_SYSTEM_EXT)))
+	$(hide) echo "avb_vbmeta_system_ext=$(BOARD_AVB_VBMETA_SYSTEM_EXT)" >> $@
+	$(hide) echo "avb_vbmeta_system_ext_args=$(BOARD_AVB_MAKE_VBMETA_SYSTEM_EXT_IMAGE_ARGS)" >> $@
+	$(hide) echo "avb_vbmeta_system_ext_key_path=$(BOARD_AVB_VBMETA_SYSTEM_EXT_KEY_PATH)" >> $@
+	$(hide) echo "avb_vbmeta_system_ext_algorithm=$(BOARD_AVB_VBMETA_SYSTEM_EXT_ALGORITHM)" >> $@
+	$(hide) echo "avb_vbmeta_system_ext_rollback_index_location=$(BOARD_AVB_VBMETA_SYSTEM_EXT_ROLLBACK_INDEX_LOCATION)" >> $@
+endif # BOARD_AVB_VBMETA_SYSTEM_EXT
+ifneq (,$(strip $(BOARD_AVB_VBMETA_PRODUCT)))
+	$(hide) echo "avb_vbmeta_product=$(BOARD_AVB_VBMETA_PRODUCT)" >> $@
+	$(hide) echo "avb_vbmeta_product_args=$(BOARD_AVB_MAKE_VBMETA_PRODUCT_IMAGE_ARGS)" >> $@
+	$(hide) echo "avb_vbmeta_product_key_path=$(BOARD_AVB_VBMETA_PRODUCT_KEY_PATH)" >> $@
+	$(hide) echo "avb_vbmeta_product_algorithm=$(BOARD_AVB_VBMETA_PRODUCT_ALGORITHM)" >> $@
+	$(hide) echo "avb_vbmeta_product_rollback_index_location=$(BOARD_AVB_VBMETA_PRODUCT_ROLLBACK_INDEX_LOCATION)" >> $@
+endif # BOARD_AVB_VBMETA_PRODUCT
 ifneq (,$(strip $(BOARD_AVB_VBMETA_VENDOR)))
 	$(hide) echo "avb_vbmeta_vendor=$(BOARD_AVB_VBMETA_VENDOR)" >> $@
 	$(hide) echo "avb_vbmeta_vendor_args=$(BOARD_AVB_MAKE_VBMETA_SYSTEM_IMAGE_ARGS)" >> $@
